ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PartyAllocationCommitter.scala

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.committer
+
+import com.codahale.metrics.Counter
+import com.daml.ledger.participant.state.kvutils.Conversions.buildTimestamp
+import com.daml.ledger.participant.state.kvutils.DamlKvutils._
+import com.daml.ledger.participant.state.kvutils.Pretty
+import com.digitalasset.daml.lf.data.Ref
+
+private[kvutils] case object PartyAllocationCommitter
+    extends Committer[DamlPartyAllocationEntry, DamlPartyAllocationEntry.Builder] {
+
+  private object Metrics {
+    // kvutils.PartyAllocationCommitter.*
+    val accepts: Counter = metricsRegistry.counter(metricsName("accepts"))
+    val rejections: Counter = metricsRegistry.counter(metricsName("rejections"))
+  }
+
+  private def rejectionTraceLog(
+      msg: String,
+      ctx: CommitContext,
+      partyAllocationEntry: DamlPartyAllocationEntry.Builder): Unit =
+    logger.trace(
+      s"Party allocation rejected, $msg, entryId=${Pretty.prettyEntryId(ctx.getEntryId)}, submId=${partyAllocationEntry.getSubmissionId}")
+
+  private val authorizeSubmission: Step = (ctx, partyAllocationEntry) => {
+    if (ctx.getParticipantId == partyAllocationEntry.getParticipantId)
+      StepContinue(partyAllocationEntry)
+    else {
+      val msg =
+        s"participant id ${partyAllocationEntry.getParticipantId} did not match authenticated participant id ${ctx.getParticipantId}"
+      rejectionTraceLog(msg, ctx, partyAllocationEntry)
+      StepStop(
+        buildRejectionLogEntry(
+          ctx,
+          partyAllocationEntry,
+          _.setParticipantNotAuthorized(
+            DamlPartyAllocationRejectionEntry.ParticipantNotAuthorized.newBuilder
+              .setDetails(msg))))
+    }
+  }
+
+  private val validateParty: Step = (ctx, partyAllocationEntry) => {
+    val party = partyAllocationEntry.getParty
+    if (Ref.Party.fromString(party).isRight)
+      StepContinue(partyAllocationEntry)
+    else {
+      val msg = s"party string '${party}' invalid"
+      rejectionTraceLog(msg, ctx, partyAllocationEntry)
+      StepStop(
+        buildRejectionLogEntry(
+          ctx,
+          partyAllocationEntry,
+          _.setInvalidName(DamlPartyAllocationRejectionEntry.InvalidName.newBuilder
+            .setDetails(msg))))
+    }
+  }
+
+  private val deduplicate: Step = (ctx, partyAllocationEntry) => {
+    val party = partyAllocationEntry.getParty
+    val partyKey = DamlStateKey.newBuilder.setParty(party).build
+    if (ctx.get(partyKey).isEmpty)
```

Comment:
here the implementation goes wrong: if `partyKey` isn't listed as an input we cannot distinguish between it actually being unallocated and submitter cheating. I will fix this here and in PackageCommitter.